cmd/soroban-rpc/internal/db/event.go

```diff
@@ -0,0 +1,336 @@
+package db
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	sq "github.com/Masterminds/squirrel"
+
+	"github.com/stellar/go/ingest"
+	"github.com/stellar/go/support/db"
+	"github.com/stellar/go/support/log"
+	"github.com/stellar/go/xdr"
+)
+
+const (
+	eventTableName = "events"
+	firstLedger    = uint32(2)
+	MinTopicCount  = 1
+	MaxTopicCount  = 4
+)
+
+// EventWriter is used during ingestion of events from LCM to DB
+type EventWriter interface {
+	InsertEvents(lcm xdr.LedgerCloseMeta) error
+}
+
+// EventReader has all the public methods to fetch events from DB
+type EventReader interface {
+	GetEvents(
+		ctx context.Context,
+		cursorRange CursorRange,
+		contractIDs [][]byte,
+		topics [][][]byte,
+		eventTypes []int,
+		f ScanFunction,
+	) error
+}
+
+type eventHandler struct {
+	log        *log.Entry
+	db         db.SessionInterface
+	stmtCache  *sq.StmtCache
+	passphrase string
+}
+
+func NewEventReader(log *log.Entry, db db.SessionInterface, passphrase string) EventReader {
+	return &eventHandler{log: log, db: db, passphrase: passphrase}
+}
+
+//nolint:gocognit,funlen,cyclop
+func (eventHandler *eventHandler) InsertEvents(lcm xdr.LedgerCloseMeta) error {
+	txCount := lcm.CountTransactions()
+
+	if eventHandler.stmtCache == nil {
+		return errors.New("EventWriter incorrectly initialized without stmtCache")
+	} else if txCount == 0 {
+		return nil
+	}
+
+	var txReader *ingest.LedgerTransactionReader
+	txReader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(eventHandler.passphrase, lcm)
+	if err != nil {
+		return fmt.Errorf(
+			"failed to open transaction reader for ledger %d: %w ",
+			lcm.LedgerSequence(), err)
+	}
+	defer func() {
+		closeErr := txReader.Close()
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	for {
+		var tx ingest.LedgerTransaction
+		tx, err = txReader.Read()
+		if errors.Is(err, io.EOF) {
+			err = nil
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		if !tx.Result.Successful() {
+			continue
+		}
```

Comment:
Don't failed transactions still have events? Or is the idea that we don't want to store those (which makes sense, just double-checking).

Reply:
Yep we do not want to store it and consistent with existing in-memory store